core/src/main/scala/kafka/tools/MirrorMaker.scala

```diff
@@ -278,9 +279,18 @@ object MirrorMaker extends Logging with KafkaMetricsGroup {
           // we only call wakeup() once to close the consumer,
           // so if we catch it in commit we can safely retry
           // and re-throw to break the loop
-          consumerWrapper.commit()
+          CoreUtils.swallow(consumerWrapper.commit(), this)
           throw e
 
+        case _: TimeoutException =>
```

Comment:
If there is performance issue and request queue time is more than 60 seconds, previously MM will keep waiting until offset commit succeeds. Now the MM just timeout offset commit operation and continue consume/produce messages. This may cause data duplication if `commitOffsets(...)` is called in `onPartitionsRevoked()` or if it is called when the MM is shutdown.

Would it make sense to change the default configuration of consumers in MM so that `default.api.timeout.ms` is set to MAX_LONG to keep old behavior as default behavior?

Going further along this line, in the scenario that offset commit can not finish in 60 seconds in `onPartitionsRevoked()`, what is the benefit to let MM give up waiting and give the partition ownership to another consumer, instead of keep trying/waiting infinitely?

Reply:
Good question. I would think this PR does not guarantee the absence of duplicate messages. Two possible points for this PR are: 1. Explicitly capturing `TimeoutException` and warning give user a hint that they might need to adjust the value of `default.api.timeout.ms`; 2. Have MM able to remove offsets for already-deleted topics from `offsets`. 

Due to the reason of Point #1 above, I do think setting `default.api.timeout.ms` to MAX_LONG is reasonable since the introduction of it does give users the ability to get rid of the indefinitely waiting. 

Does it make any sense?

Reply:
Hey @huxihx, just to clarify, you agree that we should set `default.api.timeout.ms` to MAX_LONG in MM, right?

Reply:
BTW, if the `default.api.timeout.ms` is set to MAX_LONG and a topic is deleted, what is the behavior of offset commit?

Reply:
@lindong28 Thanks for the review. Firstly, I don't quite agree to set `default.api.timeout.ms` to MAX since it's sort of against the intention why we introduce it. Secondly, the consumer will complain `Offset commit failed on partition test-0 at offset....` forever if `default.api.timeout.ms` is set to MAX and topic is deleted.

Reply:
The motivation of introducing `default.api.timeout.ms` is that user may want to avoid infinite blocking behavior. And I agree that MM should not block forever on offset commit when topic is deleted. The question here is that, if MM offset commit takes more than 60 seconds due to reasons other than topic deletion (e.g. long queue time in broker), should MM keep retrying offset commit?

It seems that if long request queue time causes offset commit in `onPartitionsRevoked ()` to timeout , most likely the next MM who owns the partition will also fail to commit offset, and in the end MM will just keep consuming/producing without committing offset which can cause issue. This seems to suggest that MM should just keep retrying offset commit in this scenario. What do you think?

Reply:
User could set it to a lower value when using MM to avoid long blocking. If no topics were deleted but TimeoutException was still thrown due to some network problem, the PR warns the user and has MM give up this round of offset commit. Although I agree that it might introduce duplicate messages when rebalance is done, MM does not make a promise to get rid of it itself. Seems we cannot expect to eliminate duplicates only via changing offset-commit behavior. Does it make sense?

Reply:
Regarding `If no topics were deleted but TimeoutException was still thrown due to some network problem`, specifically if it happens in `onPartitionsRevoked()`, could you explain a bit more why user would want MM to give up instead of keep retrying?

Reply:
Well, it's hard to say... It depends on which part user prefers. If user does not care about the duplicates, he/she will expect the MM should complete the rebalance asap, then a quick evac from committing is expected.  Assume the network error is not transient and `default.api.timeout.ms` is set to MAX, the onPartitionsRevoked will never end. In this case keeping retrying offers little help. Right?

Reply:
Just to clarify, I am not suggesting that we need to set default.api.timeout.ms to MAX_LONG. I am currently thinking that we may want to use the 60 seconds as timeout value but additionally keep retrying offset commit in e.g. onPartitionsRevoked(), and if MM can not finish offset commit after certain amount of retries, MM should fail fast.

We can discuss based on the reason of offset commit timeout. If the offset commit timeout is due to long request queue time in broker, then even if rebalance completes, the next broker will also not able to commit offset and finally all consumed data will be duplicated. In this case it seems reasonable to just let MM fail fast and let SRE investigate the performance issue in broker and possibly increase the default.api.timeout.ms value.

If offset commit failed due to persistent network error in the given MM host, regardless of whether this MM host timeout or keep retrying, the behavior of this MM host should not affect behavior of other MM. And since this MM will not be able to communicate with other hosts, we would like this MM to fail fast, which will be done by the alternative approach after certain amount of retries.

Reply:
I have to admit that I cannot figure out a better solution than yours... How about we separate the commit in `onPartitionsRevoked`  from others. That's to say, only fail fast the commit in onPartitionsRevoked but keep the rest unchanged?

Reply:
Hey @huxihx, thanks for all the work! Do you think it is reasonable to retry offset commit infinitely by default in MM per explanation above? I am happy to discuss more.